### PR TITLE
feat(gateway): Gateway.FastDirIndexThreshold

### PR DESCRIPTION
**Pull request by schomatis**
>  Toward https://github.com/ipfs/go-ipfs/issues/8178, 

TLDR: this PR skips costly  "size" column if directory is bigger or equal `Gateway.FastDirIndexThreshold`.  

Details  in https://github.com/ipfs/go-ipfs/pull/8853/files#pullrequestreview-942930258

## Demo 

You can test with `bafybeiggvykl7skb2ndlmacg2k5modvudocffxjesexlod2pfvg5yhwrqm` (has 10k items, loads fast)

>  Screen Shot 2022-04-14 at 23 54 55

---- 

<details>
  <summary>Click to expand old notes by schomatis</summary>
 
[re] "band-aid approach" from https://github.com/ipfs/go-ipfs/issues/8178#issuecomment-1088826476

```bash
ipfs config --json Gateway.FastDirIndexThreshold 3
ipfs daemon --offline

mkdir big-dir
touch big-dir/file{1..5}
BIG_DIR=$(ipfs add big-dir -r -Q)
echo "http://localhost:8080/ipfs/$BIG_DIR"
# http://localhost:8080/ipfs/QmZsx9Brmf8KGgr7o8P895WPJ7DSqTY8h2fWkvHDSnRJkM

rm big-dir/file{3..5}
BIG_DIR=$(ipfs add big-dir -r -Q)
echo "http://localhost:8080/ipfs/$BIG_DIR"
# http://localhost:8080/ipfs/QmTb1wneppbtKYuR6EwJdnieGVyrtXz3VebYU2WG3GQ1Pz
```

Take notice that there may be some caching mechanism that won't apply this limit if the directory has already been served. Take it for a spin and see if this works for your use case. It's the basic implementation of the stopgap suggested but we might need to add either a timeout or fetch entries metadata in parallel (right now this is sequential and a missing/slow entry consumes time from the total gateway timeout).

</details>

### CODE REVIEW

**Comment by schomatis** [comment]
With the current interfaces it seems browsing through the directory to get its number of entries will always be blocking on the entry node fetch. At this time asserting that a directory has, say, 5000 entries to stop its listing implies fetching 5000 nodes (defeating the purpose of the cutoff), which even if parallelizing will quickly be halted on the missing/slow entries.

Ideally I would need to extend the Directory [interface](https://github.com/ipfs/go-ipfs-files/blob/9b481f40778c3b1e65a229248c7f0b730ef72b1e/file.go#L62-L83) to be able to get its number of entries (not the entries themselves) or maybe provide another version of its [iterator](https://github.com/ipfs/go-ipfs-files/blob/9b481f40778c3b1e65a229248c7f0b730ef72b1e/file.go#L44-L60) that advances through the entries without fetching them (unless some property like Size is requested).

**Comment by schomatis** [comment]
On a second-third thought: the best I can do to deliver something useful here (which this is not it) without breaking interfaces is to make an independent `UnixfsAPI.Ls()` call to have the directory entries as fast as possible and decide if we want to go ahead and sequentially fetch entries metadata later (with the N limit of the band-aid approach). `UnixfsAPI.Ls()` is what `ipfs ls` uses and basically we can do a `ipfs ls --size=false --resolve-type=false` internally to get the parallel entry count.

Eventually we will need to make the directory iterator (linked above) parallel or switch entirely to the `UnixfsAPI.Ls()` API (which sounds like the right thing to do; we probably shouldn't have been using  `UnixfsAPI.Get()` in the first place).

I'm removing review requests and marking this as 'In Progress' again but will at least need a confirmatory :+1: here to proceed, and also will be prioritizing other issues next week (unless explicitly told otherwise, your call @lidel; this should still be an easy patch that won't take too much time).

File: config/gateway.go
Comment (lidel) [inline config/gateway.go:60]: Let's make this optional in the config, most users will be fine with implicit default.

```suggestion
	MaxDirectorySize *OptionalInteger `json:",omitempty"`
```

**Review by lidel — COMMENTED** [review]
@schomatis refactoring generated dir listing  to use `UnixfsAPI.Ls()`  is  a good plan  :+1:  
If you are ok with tacking this before CIDv1, that would be the best (I'd like to include this fix in go-ipfs 0.13, if possible) :pray:

**Review by Mitumaru — APPROVED** [review]

File: config/gateway.go
Comment (schomatis) [reply config/gateway.go:60]: Nice, didn't know this existed.

Comment (schomatis) [reply config/gateway.go]: I'm not sure `OptionalInteger` added anything different than ``json:",omitempty"`` if we're using the default Go value of 0. @lidel Could you confirm how did you envision us using this? I'm basically calling it like `MaxDirectorySize: int(cfg.Gateway.MaxDirectorySize.WithDefault(0)),`.

File: core/corehttp/gateway.go
Comment (lidel) [inline core/corehttp/gateway.go:96]:
```suggestion
			MaxDirectorySize: int(cfg.Gateway.MaxDirectorySize.WithDefault(2500)),
```

**Comment by schomatis** [comment]
@lidel Implemented a usable limit using `UnixfsAPI.Ls()`.

File: core/corehttp/gateway_handler_unixfs.go
Comment (lidel) [inline core/corehttp/gateway_handler_unixfs.go:24]: nit: we keep the context name in other places, lets to the same here
```suggestion
	// Decouple (potential) directory context to cancel it in case it's too big.
	ctx, cancel := context.WithCancel(ctx)
```

File: core/corehttp/gateway_handler_unixfs.go
Comment (lidel) [inline core/corehttp/gateway_handler_unixfs.go:79]: This should be moved to a separate func in  gateway_handler_unixfs_dir.go

File: config/gateway.go
Comment (lidel) [inline config/gateway.go:59]:
```suggestion
	// metadata. The value of 0 means no limit will be applied.
```

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (lidel) [inline core/corehttp/gateway_handler_unixfs_dir.go:141]:
```suggestion
			warnMaxDirectorySize = fmt.Sprintf("Directories bigger than %d items can't be rendered fully as the cost of generating HTML would be too big. To see all CIDs in this directory, use the CLI: ipfs ls -s --size=false --resolve-type=false %s",
```

File: config/gateway.go
Comment (lidel) [reply config/gateway.go]: We will be using it with value other than  0, something like `.WithDefault(2500)`. 
That way we can adjust it in future releases without the need for writing config migration.

File: core/coreapi/unixfs.go
Comment (lidel) [inline core/coreapi/unixfs.go:305]: @schomatis why you changed this?

File: core/corehttp/gateway.go
Comment (lidel) [inline core/corehttp/gateway.go:96]: @alanshaw is 2500 a good default limit, or should we increase it?

**Review by lidel — APPROVED** [review]
Thank you @schomatis

I took it for a spin and worked as expected, but realized the UX was not the best – it was a hard cut-off, and it is better to show some items to users.  Played a bit with different ways of communicating this and ended up with bit simpler solution:

- we list up to `Gateway.HTMLDirListingLimit`  items
- if limit is reached or surpassed, we dont fetch any remaining blocks, only show an actionable message in header and  footer 
- rebased this on top of https://github.com/ipfs/go-ipfs/pull/8872 (which removed the bindata from assets, making our code way  simpler)

Lmk your thoughts on this.

## Demo

```console
$ ipfs config --json Gateway.HTMLDirListingLimit 2
```

Opening `/ipfs/bafybeihfg3d7rdltd43u3tfvncx7n5loqofbsobojcadtmokrljfthuc7y` 

> Screen Shot 2022-04-12 at 04 13 04

Confirmed we only fetch small number of blocks by opening dir listing on an empty repo, and then shutting down the daemon. Then, in offline mode, you can see only parent and the first few blocks were fetched:

```console
$  ipfs ls -s --size=false --resolve-type=false bafybeihfg3d7rdltd43u3tfvncx7n5loqofbsobojcadtmokrljfthuc7y | wc -l
1864
$ ipfs ls -s bafybeihfg3d7rdltd43u3tfvncx7n5loqofbsobojcadtmokrljfthuc7y | wc -l
3
$ ipfs ls -s bafybeihfg3d7rdltd43u3tfvncx7n5loqofbsobojcadtmokrljfthuc7y
QmbQDovX7wRe9ek7u6QXe9zgCXkTzoUSsTFJEkrYV1HrVR -         1 - Barrel - Part 1/
QmdC5Hav9zdn2iS75reafXBq1PH4EnqUmoxwoxkS5QtuME -         10 - Pi Equals/
QmcyyLvDzCrduuvGVUQEh1DzFvM7UWGfc9sUg87PjjYCw7 -         100 - Family Circus/
Error: ipld: could not find Qmd8NDeJhzf614FSBxZwu4QD2Az14tQtJhQXJf8h4fqiSx
```

**Comment by lidel** [comment]
Note from Stewards sync: 
- change this to always list all items (no  cut-off) because we already have the root block
  - back to `Ls()` code we had before
- when item list is bigger than the limit,  skip size / type check + display header/footer informing the dir was too big to fetch details for all children

File: core/coreapi/unixfs.go
Comment (schomatis) [reply core/coreapi/unixfs.go:305]: Performance. Otherwise the listing is stalled on the consumer. (The default is arbitrary but seems sane.)

File: core/corehttp/gateway.go
Comment (schomatis) [reply core/corehttp/gateway.go:96]: Just to make sure this is the expected UX: if the user has no config set on this, it will silently use this hard-coded number. (Not opposed, just checking.)

**Comment by schomatis** [comment]
From Discord [thread](https://discord.com/channels/806902334369824788/963563468617814126/963563471352500274): we're partially moving back to the `Ls` approach since

> change this to always list all items (no cut-off) because we already have the root block

doesn't apply for the HAMT implementation (normally used for directories with many entries as in this use case).

File: core/corehttp/gateway.go
Comment (lidel) [inline core/corehttp/gateway.go:96]: :information_source: :question:  
 I've set the default to  100 to decrease load on public gateways, but lmk if should be bumped to something higher.

I picked it based on the worst case scenario where all 100 child nodes have 256KiB root blocs, which  produces  ~25MiB of additional data to be fetched by the node before it can produce HTML response.

File: docs/config.md
Comment (lidel) [inline docs/config.md:665]: :pray: Any way to make this sound less awkward?

**Review by lidel — COMMENTED** [review]
I did the refactor and pushed the updated version:

- Switched to using inexpensive dir listing via `Ls()` with `ResolveChildren(false)`. 
- Sizes from child nodes are read (fetched) only if the directory is smaller than `Gateway.FastDirIndexThreshold`
- Size column is not present in directories over `Gateway.FastDirIndexThreshold`.
- I removed footer and header because they just felt like noise. 
- Setting `Gateway.FastDirIndexThreshold` to `0` will always produce fast responses without fetching any child nodes.

Feedback appreciated.

File: docs/config.md
Comment (Winterhuman) [reply docs/config.md:665]: Maybe something like...

"The minimum number of items in a directory before the Gateway switches to a shallow, faster listing which only considers the root-most items."

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (Jorropo) [inline core/corehttp/gateway_handler_unixfs_dir.go:135]: Append with known sizes, gc aint that smart *yet*, prealoc pls.

Comment (lidel) [reply core/corehttp/gateway_handler_unixfs_dir.go:136]: @Jorropo is https://github.com/ipfs/go-ipfs/pull/8853/commits/13ba0f7d7ccb78b77b420465e012383df8fd416e what you had in mind?

File: docs/config.md
Comment (lidel) [reply docs/config.md]: Simplified in https://github.com/ipfs/go-ipfs/pull/8853/commits/736d6de0a16b96343045f69ab4a4cb1865ed2b07

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (Jorropo) [reply core/corehttp/gateway_handler_unixfs_dir.go:136]: I was thinking only of `dirListing` but that better. :)

I should push a refactor only PR that updates thoses `prealloc` find.

File: core/corehttp/gateway_handler.go
Comment (Jorropo) [inline core/corehttp/gateway_handler.go:770]: Actually here it doesn't make sense to append as this append is conditional:
```go
if root == "" {
  continue
}
```
In thoses cases, trust append.
```suggestion
	var pathRoots []string
	pathSegments := strings.Split(contentPath[6:], "/")
```

**Review by Jorropo — DISMISSED** [review]

File: config/gateway.go
Comment (alanshaw) [inline config/gateway.go:56]:
```suggestion
	// FastDirIndexThreshold is the maximum number of items in a directory
```

File: config/gateway.go
Comment (alanshaw) [inline config/gateway.go:61]: What is the default? What if it's empty?

File: docs/config.md
Comment (alanshaw) [inline docs/config.md:654]:
```suggestion
The maximum number of items in a directory before the Gateway switches
```

File: docs/config.md
Comment (alanshaw) [inline docs/config.md:657]:
```suggestion
This allows for fast listings of big directories, without the linear slowdown caused
```

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (alanshaw) [inline core/corehttp/gateway_handler_unixfs_dir.go:130]: For consistency with default size value set below for items we fetch the size for?

```suggestion
			Size:      "?", // unknown size because we did not fetch child nodes or node does not support size
```

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (alanshaw) [inline core/corehttp/gateway_handler_unixfs_dir.go:141]: Why not _always_ fetch size for the first `FastDirIndexThreshold` items so that even big directories get size information for some of their items?

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (lidel) [reply core/corehttp/gateway_handler_unixfs_dir.go:130]: I removed "?" on purpose: a wall of "?" does not improve UX in any way. 

The "?" makes sense only when dag-pb directory points at non-unixfs CID that does not have size info.

Comment (Jorropo) [reply core/corehttp/gateway_handler_unixfs_dir.go:130]: Why are we fetching sizes in the first place ?

It costy and that info I personally don't care. Can't just nuke this code ? Do some people care about that ?

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (lidel) [reply core/corehttp/gateway_handler_unixfs_dir.go:141]: Speed and simplicity: 
- Size column is either present or not. No confusing "?" after some threshold. 
- People are not tempted to naively parse HTML response to get item sizes.

File: core/corehttp/gateway_handler_unixfs_dir.go
Comment (lidel) [reply core/corehttp/gateway_handler_unixfs_dir.go:130]: Usually it is useless data, but sometimes size column allows user to reason how long the download will take: https://dist.ipfs.io/go-ipfs/v0.12.2

Personally, I'd like to stop fetching sizes by default too, but we should add an opt-in way for fetching them, if user really cares. In my mind that is something we could discuss after this non-invasive PR is merged.

**Comment by BigLep** [comment]
2022-04-21: per verbal, @Jorropo is going to do a quick check and approve/merge.

File: core/corehttp/gateway.go
Comment (Jorropo) [reply core/corehttp/gateway.go:96]: Personal opinion: should be 0. But for lack of any better solution. That fine.

**Review by Jorropo — APPROVED** [review]
LGTM

**Pull request merged**
